### PR TITLE
Don't pop-up notifications after network switch

### LINKED ISSUE #4049: After switching networks balance notifications pop up
As per @ngotchac in #4043

Also, another comment : after switching network, you get notifications about new ETH or tokens in your account (since the balances were previously at 0). Might want to handle that in another PR/Issue

### PR DESCRIPTION
**Pull request by ngotchac**
Closes #4049 

Also close notifications when clicked, and focus on the UI window/tab.
Better network switching handling.

### CODE REVIEW

**Comment by derhuerst** [comment]
There's duplicated logic with [`chainMiddleware.js`](https://github.com/ethcore/parity/blob/6ab2360637af2865ef287d30b23af8cf17156d5b/js/src/redux/providers/chainMiddleware.js#L25-L32) now. Would be nice to get this unified into a `newChain` action.

**Comment by ngotchac** [comment]
@derhuerst Ok will try to hook up there after #4077 gets merged

**Comment by ngotchac** [comment]
Hooked up to the chain middleware now.

Rewrite a part of the status and connection logic. Better handling of the disconnect and connect events (don't poll anymore when disconnected, but wait for the api transport layer to emit an `open` event).

File: js/src/api/transport/ws/ws.js [hunk 1]
```diff
@@ -131,6 +131,8 @@ export default class Ws extends JsonRpcBase {
     this._autoConnect = true;
     this._retries = 0;
 
+    this.emit('open');
```
Comment (jacogr) [inline js/src/api/transport/ws/ws.js]: Only ws? We have 2 transports, needs to be in-sync.

Additionally, since it is done on the API layer, we need to have tests that prove that is does what it is supposed to for all the transports.

Comment (ngotchac) [inline js/src/api/transport/ws/ws.js]: Well the HTTP layer will never close nor open right?

File: js/src/redux/providers/balancesActions.js [hunk 2]
```diff
@@ -65,12 +65,12 @@ function setBalances (_balances) {
...
 
           // If received a token/eth (old value < new value), notify
-          if (oldValue.lt(value) && accounts[address]) {
+          if (oldValue.lt(value) && accounts[address] && !skipNotifications) {
             const account = accounts[address];
             const txValue = value.minus(oldValue);
 
             const redirectToAccount = () => {
-              const route = `/account/${account.address}`;
+              const route = `/accounts/${account.address}`;
```
Comment (jacogr) [inline js/src/redux/providers/balancesActions.js]: Picked this up last night as well. I guess we actually need to externalise the route definitions somehow so we don't miss these again in the future.

File: js/src/redux/providers/status.js [hunk 3]
```diff
@@ -27,20 +27,49 @@ export default class Status {
...
+    });
   }
 
   start () {
+    this.stop();
+
     this._subscribeBlockNumber();
-    this._pollStatus();
-    this._pollLongStatus();
+    this._pollLongStatus(true);
```
Comment (jacogr) [inline js/src/redux/providers/status.js]: In theory (before I run this), we should now detect the chain change basically immediately? (Instead of the long delay as it is in current master.)

File: js/src/redux/providers/status.js [hunk 4]
```diff
@@ -86,10 +118,25 @@ export default class Status {
...
+    };
+
+    const { isConnected } = this._api;
+    const apiStatus = this.getApiStatus();
+
     const gotConnected = !this._apiStatus.isConnected && apiStatus.isConnected;
 
     if (gotConnected) {
-      this._pollLongStatus();
+      this._pollLongStatus(gotConnected);
```
Comment (jacogr) [inline js/src/redux/providers/status.js]: Tiny grumble - not convinced I like the `got` naming here. Would prefer a `is` or `has` especially since this one had me spinning a bit (just not well-known/consistent)

File: js/src/redux/providers/status.js [hunk 5]
```diff
@@ -173,23 +220,23 @@ export default class Status {
    * fetched every 30s just in case, and whenever
    * the client got reconnected.
    */
-  _pollLongStatus = () => {
+  _pollLongStatus = (gotConnected = false) => {
```
Comment (jacogr) [inline js/src/redux/providers/status.js]: Same here with the `got` naming.

File: js/src/redux/providers/balancesActions.js (hunk 2 above)
Comment (ngotchac) [inline js/src/redux/providers/balancesActions.js]: Might be worth it yes. Note that this route is still working

File: js/src/redux/providers/status.js (hunk 3 above)
Comment (ngotchac) [inline js/src/redux/providers/status.js]: Yep it should be immediate, and the `connecting` overlay should stay until the node is really ready.

File: js/src/secureApi.js [hunk 6]
```diff
@@ -40,6 +42,10 @@ export default class SecureApi extends Api {
     this._tryNextToken();
   }
 
+  get transport () {
```
Comment (jacogr) [inline js/src/secureApi.js]: Is there any reason we are re-defining this here? Api has both `this._transport` as well as `get transport` already and SecureApi only extends Api.

Comment (ngotchac) [inline js/src/secureApi.js]: Nope, will delete

File: js/src/api/transport/ws/ws.js (hunk 1 above)
Comment (jacogr) [inline js/src/api/transport/ws/ws.js]: It still have the isConnected & connected property. Basically when connected is set to true it should emit open, when set to false (errors), it should be emit false.

**Comment by ngotchac** [comment]
Added events to HTTP and the corresponding tests

File: js/src/api/transport/http/http.js [hunk 7]
```diff
@@ -46,19 +46,33 @@ export default class Http extends JsonRpcBase {
     };
   }
 
+  _setConnected () {
```
Comment (jacogr) [inline js/src/api/transport/http/http.js]: I think these should have been in the JsonRpc base - so both implementations are consistent with the flags they set and the events they emit (at least the base flags, not the internal extensions).

**Comment by ngotchac** [comment]
Updated the Secure API logic ! Would be great to have a second or third pair of eyes on this one ; just to make sure everything is working fine @derhuerst @jacogr 
Loggings have been added for the signer (see Settings > Parity in dev mode)

File: js/src/config.js [hunk 8]
```diff
@@ -20,6 +20,10 @@ export const LOG_KEYS = {
   TransferModalStore: {
     path: 'modals/Transfer/store',
     desc: 'Transfer Modal MobX Store'
+  },
+  Signer: {
+    path: 'secuerApi',
```
Comment (jacogr) [inline js/src/config.js]: `secureApi` - would also actually like consistency, i.e. path to actual location/entrypoint as per above entry.)

**Comment by jacogr** [comment]
Not terribly excited by the sweeping connection changes, it seems to bite us after a couple of weeks each time around. (Despite best-efforts at testing all the options - these types of sweeping & critical changes are best left for something that is not the end of a release.)

Have you been through all the connection types & tests again? (Will have to do so myself now.)

File: js/src/secureApi.js [hunk 9]
```diff
@@ -17,12 +17,15 @@
...
 import Api from './api';
+import { LOG_KEYS, getLogger } from '~/config';
 
+const log = getLogger(LOG_KEYS.Signer);
 const sysuiToken = window.localStorage.getItem('sysuiToken');
 
 export default class SecureApi extends Api {
   constructor (url, nextToken) {
-    super(new Api.Transport.Ws(url, sysuiToken, false));
+    const transport = new Api.Transport.Ws(url, sysuiToken, false);
```
Comment (jacogr) [inline js/src/secureApi.js]: Change needed?

File: js/src/secureApi.js [hunk 10]
```diff
@@ -37,11 +40,49 @@ export default class SecureApi extends Api {
       .filter((token) => token)
       .map((token) => ({ value: token, tried: false }));
 
-    this._tryNextToken();
+    this.connect();
```
Comment (jacogr) [inline js/src/secureApi.js]: Unsure why this change ended up in this PR and how it is related to the pop-up notifications. (Trying to keep things simple, the rest seems fine, however this change has now opened up a lot of seemingly unrelated tests)

File: js/src/util/notifications.js [hunk 11]
```diff
@@ -33,13 +32,34 @@ export function notifyTransaction (account, token, _value, onClick) {
...
-      x32: icon
-    },
-    timeout: 20000,
-    onClick: onClick || noop
-  });
+  let _notification = null;
+
+  Push
+    .create(`${name}`, {
+      body: `You just received ${value.toFormat()} ${token.tag.toUpperCase()}`,
```
Comment (jacogr) [inline js/src/util/notifications.js]: We are still displaying too many decimals, align this with what is in the app.

File: js/src/views/Application/Status/status.js [hunk 12]
```diff
@@ -60,6 +60,10 @@ class Status extends Component {
   renderConsensus () {
     const { upgradeStore } = this.props;
 
+    if (!upgradeStore || !upgradeStore.consensusCapability) {
```
Comment (jacogr) [inline js/src/views/Application/Status/status.js]: Not sure how these changes fit in with the rest (as notifications), apart from being formatting related.

**Comment by jacogr** [comment]
Ok, it really seems like 3 or 4 completely unrelated things are throw together under the banner. Some small, some on-point and some had me scratching my head. Just need clarification of what we are actually trying to achieve here, unless I'm missing the point of the linked issue.

File: js/src/config.js (hunk 8 above)
Comment (ngotchac) [inline js/src/config.js]: Well it is at the root of `src` so it's the right path. Wrong spelling though

File: js/src/secureApi.js (hunk 9 above)
Comment (ngotchac) [inline js/src/secureApi.js]: Not really, but doesn't hurt.

File: js/src/views/Application/Status/status.js (hunk 12 above)
Comment (ngotchac) [inline js/src/views/Application/Status/status.js]: Nothing really, but an error related to that popped up so I just fixed it

**Comment by ngotchac** [comment]
Ok so for clarifications : the PR is cleaning a bit how balances are fetched when the API is disconnected. The behavior needed for this PR to actually work (ie. not showing notifications on network change) is that nothing gets fetched when the node is disconnected, and the first fetch on reconnection should be initiated by the `chainMiddleware` that is aware of a network change and thus can fetch balances without showing notifications.

So, in order to be able to do that, the Redux `status` logic needed some change, ie. stop polling when the connection is closed. That's why I had to introduce `close` (and thus `open`) events to the API transport layer, so that we can stop polling after the `close` event is emitted.

Now, without polling all the time the API status (which is no more needed, and is actually better), the signer needs a cleanup so that it can show the proper overlay on disconnect, etc. That's why all these 3 parts of the UI have been touched.

File: js/src/secureApi.js (hunk 10 above)
Comment (ngotchac) [inline js/src/secureApi.js]: Cf. my comment below

File: js/src/secureApi.js [hunk 13]
```diff
@@ -136,51 +269,82 @@ export default class SecureApi extends Api {
...
+   * and running (it might take some time before
+   * the node is actually ready even when the client
+   * is connected).
+   *
+   * We check that the `parity_enode` RPC calls
+   * returns successfully
+   */
+  _waitUntilNodeReady () {
+    return this
+      .parity.enode()
```
Comment (jacogr) [inline js/src/secureApi.js]: My big concern with this is that I have seen cases where this never gets a valid value. Parity appears to be running, however enode is not set. (Parity core issue no doubt, may have been fixed, but still a concerns)

Comment (ngotchac) [inline js/src/secureApi.js]: Ok, true. What about a mac response time (eg. 10s) after which it's get validated anyhow...

File: js/src/secureApi.js [hunk 14]
```diff
@@ -17,40 +17,129 @@
 import { uniq } from 'lodash';
 
 import Api from './api';
+import { LOG_KEYS, getLogger } from '~/config';
 
+const log = getLogger(LOG_KEYS.Signer);
 const sysuiToken = window.localStorage.getItem('sysuiToken');
 
 export default class SecureApi extends Api {
+
```
Comment (jacogr) [inline js/src/secureApi.js]: Extra blank line.

File: js/src/secureApi.js [hunk 15]
```diff
@@ -17,40 +17,129 @@
...
+  _isConnecting = false;
+  _needsToken = false;
+  _tokens = [];
+
```
Comment (jacogr) [inline js/src/secureApi.js]: extra blank line.

File: js/src/secureApi.js (hunk 13 above)
Comment (jacogr) [inline js/src/secureApi.js]: It is a tough one, since from the UI perspective, everything works even without the enode. Well, it appears working. I suppose my preference would have been an "enode not ready" status warning, instead of waiting for it - not to be done here and now though, that check just has me slightly uncomfortable...

**Comment by ngotchac** [comment]
Finally got through with this. Had to rewrite some of the Tokens Balances fetching logic, which was a bit messy and inefficient (called multiple times on load, sometimes it would take a new block for the tokens to be shown). Now everything seems to work fine, no notifications on network switch, correct updating of the tokens. We might want to tackle a performance issue that can be seen on the main net (in another PR). We could implement a way to only fetch the tokens balances for which the UI addresses (accounts, address book, contracts) have received a `Transfer()` event (one call thanks to `eth_getLogs`).
